processing/src/main/java/io/druid/segment/DimensionMerger.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import com.google.common.io.OutputSupplier;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.Buffer;
+import java.nio.IntBuffer;
+import java.util.List;
+
+/**
+ * Processing related interface
+ *
+ * A DimensionMerger is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the segment merging process (i.e., work done by IndexMerger/IndexMergerV9).
+ *
+ * This object is responsible for:
+ *   - merging any relevant structures from the segments (e.g., encoding dictionaries)
+ *   - writing the merged column data and any merged indexing structures (e.g., dictionaries, bitmaps) to disk
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ */
+public interface DimensionMerger<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
```

Comment:
In general, this abstraction feels very leaky to me.  There are a lot of implementation details like v8 versus v9 and likely  interplay between the method calls, but it's not clear how they interplay.

For v8 versus v9, if there is a significant enough difference between how they work, then I think `DimensionHandler` needs to have two different methods: one catering to each semantic.

For the interplay of the methods, I think that this interface is likely in the wrong place.  Interfaces are generally most useful either when they are high-level (a bunch of methods that give all of the responsibility to the object, something like the `PostAggregator` would be an example) or low-level (a bunch of getter/setter type things that a driver has complete control over, the Collections and Map interfaces would be examples).

I think this interface needs to move in either one of those directions.  I'm not sure which, but reading the interface and the javadoc on it has not given me confidence that I would know what to do if I were trying to implement the interface myself.